docs/adrs/004-privacy-invariants.md

```diff
@@ -0,0 +1,64 @@
+
+### Prax Extension
+
+```
+Invariant 1.
+
+The extension must ensure that its local and session storage states are adequately protected to prevent the leakage of sensitive information to potential attackers.
+```
+
+[Chrome's Storage API](https://developer.chrome.com/docs/extensions/reference/api/storage) provides an extension specific way to persist user state and data. Particularly, the extension's service worker and content scripts have access to the storage API.
+
+- **storage.local** data is stored locally and cleared when extension is removed.
+
+- **storage.session** holds data in memory for the duration of a browser session.
+
+In https://github.com/penumbra-zone/web/pull/38, we implement secure password-based encryption and decryption using PBKDF2 and AES-GCM, as well as mnemonic phrase generation and validation.
+
+1. **PBKDF2 algorithm**: key derivation function with SHA-512 key stretching and 210,000 iterations to hamper brute force, dictionary attacks, and precomputed attacks (rainbow tables) attacks.
+2. **AES-GCM** symmetric encryption scheme, using a random nonce for each encryption operation during wallet initialization.
+3. **SHA-256** hash of the derived key material.
+
+Currently, the key material is stored only in the extension's session storage upon login and is correctly cleared when the extension is locked. This key material is used to unseal the encrypted seed phrase ciphertext stored in local state storage and generate the associated spend key.
+
+#### <ins>Remediations</ins>
+
+While this behavior is correct and expected, there are additional potential privacy-related leaks to consider:
+
+1. The full viewing key material is generated from the spend key, and the plaintext JSON of the full viewing key is stored in local state storage. When the extension is locked, a potential attacker with access to the console can still execute `chrome.storage.local.get().then(console.log)` and access the full viewing key in the clear. To mitigate this, we should:
+
+   - Ensure that all access to sensitive data in state storage is gated behind an authentication check to ensure user has logged in,
+   - Consider storing the encrypted full viewing key or not storing the full viewing key at all and deriving it from the spend key each time it is needed.
+
+2. Require a password for the extension upon mainnet launch and warn users that setting an empty password is not advised.
+3. Avoid transmitting errors by default in case keys are exposed in error data. Currently, `unseal` method for instance has the potential to transmit errors that could expose sensitive key information through the error data. Instead, handle errors securely and generalize the emitted error messages.
```

Comment:
Interesting! Have you validated that the error does leak something? Given it only throws if there is an issue, I suppose at worst it exposes the cipher text? Regardless, seems like an easy fix to obfuscate this: https://github.com/penumbra-zone/web/blob/cf63b30373d3d67f5783517fdef0e0e9f46fa918/packages/crypto/src/encryption.ts#L155 and just return a `null` if there's an error.

Reply:
I haven't validated this, but it's been known to happen in general cryptography-related applications. Thinking we could do something along these lines:

```
try {
    // Decrypt the ciphertext using the nonce and key
    return await decrypt(box.cipherText, box.nonce, this.key);
  } catch (e) {
    // Handle specific known errors without exposing sensitive information
    if (e instanceof TypeError || (e instanceof DOMException && e.name === 'OperationError')) {
      return null;
    }
    // Log a generic error message
    console.error('Decryption failed due to an unexpected error');
    return null;
  }
```